File: apps/kotlin/src/main/kotlin/chapter10/RandomForest.kt
package chapter10

import chapter03.DecisionTree
import org.jetbrains.kotlinx.dataframe.AnyFrame
import org.jetbrains.kotlinx.dataframe.api.toDataFrame
import kotlin.random.Random

fun majorityVote(votes: List<List<String>>): List<String> =
    votes.first().indices.map { sample ->
        votes
            .map { it[sample] }
            .groupingBy { it }
            .eachCount()
            .maxBy { it.value }
            .key
    }

fun bootstrapSample(
    size: Int,
    random: Random,
): List<Int> = List(size) { random.nextInt(size) }

fun AnyFrame.selectColumns(names: List<String>): AnyFrame = names.map { this[it] }.toDataFrame()

data class FittedTree(
    val columns: List<String>,
    val rows: List<Int>,
    val model: DecisionTree,
)

class RandomForest(
    private val nEstimators: Int = 10,
    private val maxFeatures: Int = 2,
    private val maxDepth: Int? = null,
    private val seed: Int = 0,
) : Classifier {
    var trees: List<FittedTree> = emptyList()
        private set

    override fun fit(
        x: AnyFrame,
        t: List<String>,
    ): RandomForest {
        val random = Random(seed)
        trees =
            List(nEstimators) {
                val rows = bootstrapSample(x.rowsCount(), random)
                val chosen =
                    x
                        .columnNames()
                        .shuffled(random)
                        .take(maxFeatures)
                        .toSet()
                val columns = x.columnNames().filter { it in chosen }
                val model = DecisionTree(maxDepth).fit(x[rows].selectColumns(columns), t.slice(rows))
                FittedTree(columns = columns, rows = rows, model = model)
            }
        return this
    }

    override fun predict(x: AnyFrame): List<String> {
        check(trees.isNotEmpty()) { "fit で学習してから predict を呼んでください" }
        return majorityVote(trees.map { it.model.predict(x.selectColumns(it.columns)) })
    }
}

File: apps/python/lib/chapter10/random_forest.py
from collections import Counter

import numpy as np
import pandas as pd
from numpy.typing import NDArray

from lib.chapter03.decision_tree import DecisionTree


def majority_vote(votes: list[list[str]]) -> list[str]:
    return [Counter(sample).most_common(1)[0][0] for sample in zip(*votes, strict=True)]


def bootstrap_sample(size: int, rng: np.random.Generator) -> NDArray[np.int64]:
    return rng.integers(0, size, size=size)


class RandomForest:
    def __init__(
        self,
        n_estimators: int = 10,
        max_features: int = 2,
        max_depth: int | None = None,
        seed: int = 0,
    ) -> None:
        self.n_estimators = n_estimators
        self.max_features = max_features
        self.max_depth = max_depth
        self.seed = seed
        self.trees: list[tuple[list[str], DecisionTree]] = []
        self.bootstrap_rows: list[NDArray[np.int64]] = []

    def fit(self, x: pd.DataFrame, t: pd.Series) -> "RandomForest":
        rng = np.random.default_rng(self.seed)
        self.trees = []
        self.bootstrap_rows = []
        for _ in range(self.n_estimators):
            rows = bootstrap_sample(len(x), rng)
            self.bootstrap_rows.append(rows)
            chosen = rng.choice(len(x.columns), size=self.max_features, replace=False)
            columns = [str(x.columns[i]) for i in sorted(chosen)]
            tree = DecisionTree(max_depth=self.max_depth)
            tree.fit(x.iloc[rows][columns], t.iloc[rows])
            self.trees.append((columns, tree))
        return self

    def predict(self, x: pd.DataFrame) -> list[str]:
        if not self.trees:
            raise ValueError("fit で学習してから predict を呼んでください")
        return majority_vote([tree.predict(x[columns]) for columns, tree in self.trees])
